### DISCUSSION TITLE
Reducing loop delay

### DISCUSSION THREAD

**Discussion by perceptualrobots** [cyberbotics/wrestling#58] [Q&A] [answered]
I was trying to reduce the time of each iteration. Here are results of some tests with different settings in the world file and the code and running the application as normal as from cmd line in fast mode.

![image](https://user-images.githubusercontent.com/67976964/225763247-573f8a5c-d6b8-4fa9-9c48-7612e36402fa.png)

Is it the case that the smallest loop time attainable is around 10 ms?

Why is the actual loop time 26 ms when run from cmd line with file setting of 5 ms but code setting of 20 ms? Whereas the last line is 10 ms.

Here is the top level code:

    ```
    wrestler = WrestlerSupervisor()    
    time_step=20
    max_loops=500
    
    wrestler.simulationReset()
    tic = time.perf_counter()
    wrestler.run(CI, time_step=time_step, max_loops=max_loops)
    toc = time.perf_counter()
    elapsed = 1000 * (toc-tic)
    print(f'Elapsed time: {elapsed:4.0f}')  
```

**Comment by omichel** [comment, marked as answer]
The basic time step is used by Webots to compute the physics. If this value is low, it will impact the speed of the simulation. I am not sure what you mean by override time step, but that means the time step used by the controller program, then all the results displayed here looks coherent to me. There is obviously a minimum real time step that depends both on the complexity of the simulation and horsepower of the computer.

**Reply by perceptualrobots** [reply]
> The basic time step is used by Webots to compute the physics. If this value is low, it will impact the speed of the simulation. I am not sure what you mean by override time step, but that means the time step used by the controller program, then all the results displayed here looks coherent to me. There is obviously a minimum real time step that depends both on the complexity of the simulation and horsepower of the computer.

The basicTimeStep column is the value that is in the file. The "override timestep" column is the value I set in the program to override the value in the file, i.e. passing in to self.step(). Is this allowed, or is the value from the file used elsewhere?

For example, in row 6 the value in the file is 5, but I set it as 20 in the code. In row 8, the value is 20 in the file and in the code.

So, why is the actual loop time taking much longer on row 6 than row 8?

![image](https://user-images.githubusercontent.com/67976964/225900859-466bee8b-645e-40ee-927a-85b5030c372a.png)

**Reply by omichel** [reply]
> Is this allowed, or is the value from the file used elsewhere?

Yes, this is allowed to pass any value to the `Robot.step()` method. However, in practice, it is optimal to pass a multiple of the `WorldInfo.basicTimeStep` value. In this competition scenario the basic time step is 20, so values like 20, 40, 60, etc. are the best to be passed to the `Robot.step()` method.

Therefore rows 3 and 7 in your example doesn't make sense because the simulation won't do anything for 20/5 = 4 steps of your controller.

The loop takes longer on row 6 because the simulator is working more intensively (performs 4 steps of 5 ms) while the controller performs only on step of 20 ms, whereas in row 8, the simulator performs only one step of simulation for one step of the controller loop, which is less accurate in terms of simulation realism but runs faster.

**Reply by perceptualrobots** [reply]
Just to follow up on this, is it the case that the simulation uses the timestep value from the file and the controller can use a different value (the one passed into Robot.step())? (I was thinking that the simulation was governed by what was passed from the controller.)

What then is the timestep if webots is run with --mode=fast?

**Reply by omichel** [reply]
The time step of Webots is specified in the WorldInfo.basicTimeStep value of the world file. In fast mode, it is a virtual time whereas in real-time mode, Webots tried to keep it real-time. The time step of the controller should be a multiple of the WorldInfo.basicTimeStep. See explanations here: https://cyberbotics.com/doc/guide/controller-programming#the-step-and-wb_robot_step-functions.
